### ISSUE TITLE
Discussion: RankSEG as optional inference-time post-processing for segmentation probabilities

### ISSUE THREAD

**Issue by Leev1s** [torchgeo/torchgeo#3768]
## Summary

I wanted to share RankSEG as a possible inference-time post-processing direction for TorchGeo segmentation workflows, in case it is useful for future discussion.

RankSEG is a training-free decision rule for probabilistic segmentation outputs. It is not a loss function, fine-tuning method, or checkpoint. For a trained segmentation model, the flow would be:

```text
image -> trained TorchGeo model -> logits -> softmax/sigmoid probabilities -> RankSEG -> mask
```

In other words, it would only affect the final decision step after probabilities have already been produced. The model, checkpoint, and logits remain unchanged. I do not want to assume that this belongs in TorchGeo. 

## Rationale

This may be relevant to TorchGeo because several segmentation workflows already expose the probability interface needed for this kind of method:

- `SemanticSegmentationTask.predict_step` can expose sigmoid or softmax probabilities, depending on the task configuration.
- TorchGeo includes segmentation datasets with ground-truth masks and IoU/F1-style metrics.
- Remote-sensing segmentation tasks often have strong class imbalance, where the final threshold or `argmax` rule can affect Dice/IoU behavior.

The intent would not be to replace the default `argmax` or thresholding behavior. Those defaults are simple, fast, and broadly understandable. The question is only whether a label-free, metric-aware inference-time decision rule is worth documenting or making easier to try for users who care about Dice/IoU behavior on imbalanced segmentation tasks.

There is also a related discussion in MONAI about RankSEG-style decoding as an optional post-processing step for medical segmentation workflows: [Project-MONAI/MONAI#8857](https://github.com/Project-MONAI/MONAI/issues/8857). I mention this only as broader context that similar probability-to-mask decoding questions may be relevant across segmentation libraries; TorchGeo's maintainers may of course make a different design decision based on geospatial use cases and maintenance scope.

## Small local check

To keep this separate from training, I ran a small frozen-checkpoint CPU sanity check using TorchGeo's pretrained Fields of The World U-Net:

- Dataset: `FieldsOfTheWorld`
- Country subset: Cambodia
- Target: `3-class`
- Ground truth: FTW background/field/field-boundary masks
- Model: `torchgeo.models.unet`
- Weights: `Unet_Weights.SENTINEL2_3CLASS_FTW`
- Training: none
- Device: CPU
- Samples: first 32 validation samples and first 32 test samples, used only as a non-random smoke check
- Compared final decision rules: `argmax`, `RankSEG(metric="iou")`, and `RankSEG(metric="dice")`

The only change was the post-processing step after `softmax`; no labels were used by RankSEG at inference time.

Metric comparison:

![FTW pretrained U-Net metric comparison](https://files.seeusercontent.com/2026/06/04/4kiR/ftw_pretrained_final_metrics.png)

Direct link: https://files.seeusercontent.com/2026/06/04/4kiR/ftw_pretrained_final_metrics.png

Qualitative examples:

![FTW pretrained U-Net qualitative examples](https://files.seeusercontent.com/2026/06/04/1tWu/ftw_pretrained_qualitative_grid.png)

Direct link: https://files.seeusercontent.com/2026/06/04/1tWu/ftw_pretrained_qualitative_grid.png

Validation results:

| method | mIoU | mDice | field IoU | field Dice | accuracy |
|---|---:|---:|---:|---:|---:|
| argmax | 0.2246 | 0.3660 | 0.2604 | 0.4132 | 0.3688 |
| RankSEG-IoU | 0.2597 | 0.4077 | 0.3542 | 0.5231 | 0.4302 |
| RankSEG-Dice | 0.2676 | 0.4161 | 0.3793 | 0.5500 | 0.4461 |

Test results:

| method | mIoU | mDice | field IoU | field Dice | accuracy |
|---|---:|---:|---:|---:|---:|
| argmax | 0.3881 | 0.5507 | 0.3831 | 0.5540 | 0.5132 |
| RankSEG-IoU | 0.4176 | 0.5810 | 0.4571 | 0.6274 | 0.5614 |
| RankSEG-Dice | 0.4250 | 0.5883 | 0.4776 | 0.6464 | 0.5743 |

## Possible ways this could fit

If maintainers think the direction is worth exploring, possible scopes might include:

1. Keep RankSEG external and add only a short recipe showing how to apply a post-processor to TorchGeo segmentation probabilities.
2. If TorchGeo already has an appropriate extension point, document how users could attach this kind of post-processor.
3. Leave TorchGeo unchanged, and keep the example in the RankSEG repository while referencing TorchGeo workflows there.

I would also like to thank @statmlben and @ZixunWang for the RankSEG method. Any mistakes in this issue draft or the small smoke test setup are mine.

**Comment by adamjstewart** [comment]
Seems cool, curious what @isaaccorley @calebrob6 think of the FTW results!

Would love to integrate this into TorchGeo, just not sure what the best approach would be. PyTorch Lightning offers callbacks, but those occur after loss/metric computation. We could add a helper function that selects between different methods. Are there other methods out there worth adding at the same time? I don't want the trainers to be too complex, but I also want the best results we can get with the least code duplication.

**Comment by isaaccorley** [comment]
Hi @Leev1s RankSeg seems super useful particular for the FTW models. We actually already have the new and improved PRUE FTW models which I'd love to redo your experiments with.

We can also consider contributing directly to the FTW repo which is used to run the Global FTW run! 

https://github.com/fieldsoftheworld/ftw-baselines

**Comment by Leev1s** [comment]
Thanks @adamjstewart and @isaaccorley for the thoughtful pointers!

On the integration side, I agree that keeping the trainers simple is important. One related data point from the Transformers side is [huggingface/transformers#46139](https://github.com/huggingface/transformers/issues/46139) and the follow-up PR [huggingface/transformers#46163](https://github.com/huggingface/transformers/pull/46163), where the discussion is about exposing semantic segmentation pre-final scores/probability-like tensors from post-processing APIs instead of only returning the final hard segmentation map. I do not think TorchGeo has the same gap in exactly the same way: `SemanticSegmentationTask.predict_step` already returns a metadata-aware dictionary with `probabilities`, `bounds`, and `transform`, and the change-detection / spatiotemporal segmentation tasks also apply `sigmoid` or `softmax` in `predict_step`. So TorchGeo already has a fairly good probability-level inference interface. The remaining question seems narrower: whether there should be a small, optional probability-to-mask decoding helper or recipe after `predict_step`.

For RankSEG specifically, the current integration pattern is intentionally narrow: keep the model/checkpoint/inference path unchanged, restore or compute probabilities, and replace only the final probability-to-mask step. The minimal PyTorch form is basically:

```python
from rankseg import RankSEG
import torch

with torch.inference_mode():
    logits = model(images)
    probs = torch.softmax(logits, dim=1)

    post = RankSEG(metric="dice", solver="RMA", output_mode="multiclass")
    preds = post(probs)
```

For one-off use, the current public API also exposes the functional form:

```python
from rankseg.functional import rankseg

preds = rankseg(probs, metric="dice", solver="RMA", output_mode="multiclass")
```

In the RankSEG repo, the Hugging Face `transformers` integration follows the same idea but wraps the model-specific probability restoration step:

```python
from rankseg.integration import transformers

preds = transformers.postprocess(
    outputs,
    target_sizes=target_sizes,
    rankseg_kwargs={"metric": "dice", "solver": "RMA"},
)
```

So if this ends up fitting TorchGeo, I would imagine a small helper or recipe around “probabilities -> postprocessor -> masks” before thinking about trainer-level changes. That would keep the current loss/metric behavior on logits unchanged, and would put RankSEG after `predict_step` as a decoder rather than as a loss, metric, or model change. It may also be worth thinking about whether the interface should be generic enough for other post-processing methods, rather than RankSEG-specific.

On the FTW side, thanks for pointing me to the PRUE models. I should say that my current FTW context is still shallow: the numbers above are only a small TorchGeo smoke test with `Unet_Weights.SENTINEL2_3CLASS_FTW`, not a proper FTW benchmark. A reasonable next step could be a small reproducible script/notebook with:

- frozen PRUE checkpoint;
- fixed FTW country/split subset;
- default argmax vs `RankSEG(metric="iou")` / `RankSEG(metric="dice")`;

If the effect still looks useful there, I am also open to moving the experiment toward `fieldsoftheworld/ftw-baselines`, especially if that is closer to the Global FTW inference pipeline.

[https://github.com/rankseg/rankseg](https://github.com/rankseg/rankseg)

**Comment by adamjstewart** [comment]
`predict_step` is quite rarely used, it's only for inference after train/val/test is complete. I think most users will want this for all 4 steps. This is a bit harder to support.

**Comment by isaaccorley** [comment]
This is meant to be a test-time (predict) method. I don't think users need this for train or val steps, maybe test if you want to use it for metrics.

**Comment by adamjstewart** [comment]
Val metrics are also important for hparam tuning.

**Comment by isaaccorley** [comment]
It's meant for improving a trained model's outputs. At early epochs it would be meaningless overhead.

**Comment by Leev1s** [comment]
The main design question is where such a decoder should be exposed in TorchGeo. It could be used only for frozen-checkpoint inference/post-processing, but it could also potentially be used for validation/test metric computation if users want Dice/IoU to be evaluated on RankSEG decoded masks rather than the default threshold/argmax-decoded masks.

**Comment by statmlben** [comment]
Thank you a lot for the discussion @adamjstewart @isaaccorley. After discussing this with our group @Leev1s @ZixunWang, we looked more closely at the current `torchgeo.trainers.segmentation` design.

If we understand correctly, the current `train_metrics` and/or `test_metrics` in `torchgeo.trainer.segmentation` are shared metrics built on top of `torchmetrics.classification`, such as `Accuracy`, `F1Score`, `JaccardIndex`, `Precision`, and `Recall`. In that setup, the **probability-to-mask** conversion seems to be handled inside the `TorchMetrics` classification metrics, only through thresholding decoding. Because of this, it may be a little difficult to integrate RankSEG cleanly into the current metric pipeline.

We also wanted to share one possible suggestion. For segmentation tasks, metrics such as F1/Dice and Jaccard/IoU may have slightly different behavior from their classification counterparts, see [torchmetrics.segmentation.DiceScore](https://lightning.ai/docs/torchmetrics/stable/segmentation/dice.html), especially when considering options such as `aggregation_level='samplewise'`. So perhaps it could be worth considering segmentation-specific metrics from `torchmetrics.segmentation` for TorchGeo segmentation tasks.

More broadly, using the simple thresholding behavior from classification metrics may be somewhat suboptimal for segmentation, since it treats each pixel independently when converting probabilities into masks. In contrast, methods such as RankSEG try to use more global information from the segmentation output during decoding/post-processing. This is one reason why having an explicit probability-to-mask decoding step could be useful for segmentation tasks.

If that direction makes sense, then the **probability-to-mask decoding** step would likely need to become an explicit part of the segmentation pipeline, since `torchmetrics.segmentation` generally expect the prediction to already be a one-hot/boolean mask. This might also make the design more flexible beyond `RankSEG`, as there can be meaningful choices in this decoding/post-processing step. For example, frameworks such as [MONAI.transforms](https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/transform.py) and [transformer](https://github.com/huggingface/transformers/blob/e86792e30d1506422987bd6bc819d7edfed33ae7/src/transformers/models/oneformer/processing_oneformer.py#L159) segmentation pipelines expose similar post-processing hooks.

  ```mermaid
  flowchart LR
      A["Model output<br/>logits / probabilities"] --> B["Mask decoder<br/>threshold / argmax / RankSEG / custom"]
      B --> C["One-hot / boolean<br/>prediction mask"]
      C --> E["torchmetrics.segmentation<br/>DiceScore / MeanIoU / etc."]
      E --> D["train_metrics / test_metrics<br/>in torchgeo.trainers.segmentation"]
      F["Target mask"] --> E
      F["Target mask"] --> G

      A -. current .-> G["torchmetrics.classification<br/>F1Score / JaccardIndex / etc."]
      G -. internal decoding .-> D
  ```

Of course, this is only a suggestion, and we are very happy to follow TorchGeo’s preferred design direction. If such a change would be useful, we would be glad to help with the discussion and implementation.
